Mathlib/CategoryTheory/Monad/Basic.lean

```diff
@@ -267,104 +253,113 @@ theorem comp_toNatTrans {T₁ T₂ T₃ : Comonad C} (f : T₁ ⟶ T₂) (g : T
+@[simp]
+def MonadIso.mk {M N : Monad C} (f : (M : C ⥤ C) ≅ N)
+    (f_η : ∀ (X : C), M.η.app X ≫ f.hom.app X = N.η.app X)
+    (f_μ : ∀ (X : C), M.μ.app X ≫ f.hom.app X =
+    (M.map (f.hom.app X) ≫ f.hom.app (N.obj X)) ≫ N.μ.app X) : M ≅ N where
+  hom :=
+    { toNatTrans := f.hom
+      app_η := f_η
+      app_μ := f_μ }
   inv :=
     { toNatTrans := f.inv
-      app_η' := fun X => by simp [← f_η]
-      app_μ' := fun X => by
-        rw [← nat_iso.cancel_nat_iso_hom_right f]
-        simp only [nat_trans.naturality, iso.inv_hom_id_app, assoc, comp_id, f_μ,
-          nat_trans.naturality_assoc, iso.inv_hom_id_app_assoc, ← functor.map_comp_assoc]
+      app_η := fun X => by simp [← f_η]
+      app_μ := fun X => by
+        rw [← NatIso.cancel_natIso_hom_right f]
+        simp only [NatTrans.naturality, Iso.inv_hom_id_app, assoc, comp_id, f_μ,
+          NatTrans.naturality_assoc, Iso.inv_hom_id_app_assoc, ← Functor.map_comp_assoc]
         simp }
 #align category_theory.monad_iso.mk CategoryTheory.MonadIso.mk
 
+-- porting note: was @[simps]
 /-- Construct a comonad isomorphism from a natural isomorphism of functors where the forward
 direction is a comonad morphism. -/
-@[simps]
-def ComonadIso.mk {M N : Comonad C} (f : (M : C ⥤ C) ≅ N) (f_ε f_δ) : M ≅ N
-    where
-  Hom :=
-    { toNatTrans := f.Hom
-      app_ε' := f_ε
-      app_δ' := f_δ }
+@[simp]
+def ComonadIso.mk {M N : Comonad C} (f : (M : C ⥤ C) ≅ N)
+    (f_ε : ∀ (X : C), f.hom.app X ≫ N.ε.app X = M.ε.app X)
+    (f_δ : ∀ (X : C), f.hom.app X ≫ N.δ.app X =
+    M.δ.app X ≫ f.hom.app (M.obj X) ≫ N.map (f.hom.app X)) : M ≅ N where
+  hom :=
+    { toNatTrans := f.hom
+      app_ε := f_ε
+      app_δ := f_δ }
   inv :=
     { toNatTrans := f.inv
-      app_ε' := fun X => by simp [← f_ε]
-      app_δ' := fun X => by
-        rw [← nat_iso.cancel_nat_iso_hom_left f]
-        simp only [reassoc_of (f_δ X), iso.hom_inv_id_app_assoc, nat_trans.naturality_assoc]
-        rw [← functor.map_comp, iso.hom_inv_id_app, Functor.map_id]
+      app_ε := fun X => by simp [← f_ε]
+      app_δ := fun X => by
+        rw [← NatIso.cancel_natIso_hom_left f]
+        simp only [reassoc_of% (f_δ X), Iso.hom_inv_id_app_assoc, NatTrans.naturality_assoc]
+        rw [← Functor.map_comp, Iso.hom_inv_id_app, Functor.map_id]
         apply (comp_id _).symm }
 #align category_theory.comonad_iso.mk CategoryTheory.ComonadIso.mk
 
 variable (C)
 
 /-- The forgetful functor from the category of monads to the category of endofunctors.
 -/
-@[simps]
+@[simps!]
 def monadToFunctor : Monad C ⥤ C ⥤ C where
   obj T := T
-  map M N f := f.toNatTrans
+  map f := f.toNatTrans
 #align category_theory.monad_to_functor CategoryTheory.monadToFunctor
 
 instance : Faithful (monadToFunctor C) where
 
-@[simp]
+/-- Porting note: removed @[simp] as the linter complains of the LHS being reducible to
+  (comonadToFunctor _).mapIso (Iso.mk (MonadHom.mk f.hom) (MonadHom.mk f.inv))
+  which does not even compile
+-/
```

Comment:
That doesn't sound very promising. If you have `set_option pp.all true`, does it reduce to something that typechecks? Should we have a new simp-lemma for that LHS, so that the total behaviour of the simp-set doesn't change?

Reply:
OK, consider `monadToFunctor_mapIso_monad_iso_mk` first. simpNF says the LHS is reducible to
```lean
CategoryTheory.Functor.mapIso (CategoryTheory.monadToFunctor C)
  (CategoryTheory.Iso.mk (CategoryTheory.MonadHom.mk f.hom) (CategoryTheory.MonadHom.mk f.inv))
```

But when I make a new theorem with this simplification
```lean
theorem monadToFunctor_mapIso_monad_iso_mk' {M N : Monad C} (f : (M : C ⥤ C) ≅ N) :
    CategoryTheory.Functor.mapIso (CategoryTheory.monadToFunctor C)
    (CategoryTheory.Iso.mk (CategoryTheory.MonadHom.mk f.hom) (CategoryTheory.MonadHom.mk f.inv)) = f :=
  by aesop_cat
```

it is not the _body_ of the proof that throws an error, but `MonadHom.mk f.hom` and `MonadHom.mk f.inv` _in the "simplified" LHS itself_!

```
tactic 'aesop' failed, failed to prove the goal after exhaustive search.
T: Monad C
G: Comonad C
MN: Monad C
f: M.toFunctor ≅ N.toFunctor
⊢ ∀ (X : C), (Monad.η M).app X ≫ f.hom.app X = (Monad.η N).app X
Basic.lean:327:28
tactic 'aesop' failed, failed to prove the goal after exhaustive search.
T: Monad C
G: Comonad C
MN: Monad C
f: M.toFunctor ≅ N.toFunctor
⊢ ∀ (X : C),
  (Monad.μ M).app X ≫ f.hom.app X = (M.toFunctor.map (f.hom.app X) ≫ f.hom.app (N.toFunctor.obj X)) ≫ (Monad.μ N).app X
Basic.lean:327:63
tactic 'aesop' failed, failed to prove the goal after exhaustive search.
T: Monad C
G: Comonad C
MN: Monad C
f: M.toFunctor ≅ N.toFunctor
⊢ ∀ (X : C), (Monad.η N).app X ≫ f.inv.app X = (Monad.η M).app X
Basic.lean:327:63
tactic 'aesop' failed, failed to prove the goal after exhaustive search.
T: Monad C
G: Comonad C
MN: Monad C
f: M.toFunctor ≅ N.toFunctor
⊢ ∀ (X : C),
  (Monad.μ N).app X ≫ f.inv.app X = (N.toFunctor.map (f.inv.app X) ≫ f.inv.app (M.toFunctor.obj X)) ≫ (Monad.μ M).app X
```

Reply:
I suggest you inspect the statement with `pp.all` enabled. That way you'll have `@`s all over the place, but it will also make clear what is actually going on.

Reply:
@jcommelin I've found that exactly four extra arguments – no more, no less – are needed for `monadToFunctor_mapIso_monad_iso_mk'` to type-check and from there compile:
```lean
@[simp]
theorem monadToFunctor_mapIso_monad_iso_mk' {M N : Monad C} (f : (M : C ⥤ C) ≅ N)
    (f_η : ∀ (X : C), M.η.app X ≫ f.hom.app X = N.η.app X)
    (f_μ : ∀ (X : C), M.μ.app X ≫ f.hom.app X =
    (M.map (f.hom.app X) ≫ f.hom.app (N.obj X)) ≫ N.μ.app X)
    (f_η' : ∀ (X : C), N.η.app X ≫ f.inv.app X = M.η.app X)
    (f_μ' : ∀ (X : C), N.μ.app X ≫ f.inv.app X =
    (N.map (f.inv.app X) ≫ f.inv.app (M.obj X)) ≫ M.μ.app X):
    (monadToFunctor C).mapIso (Iso.mk (MonadHom.mk f.hom) (MonadHom.mk f.inv)) = f := by
  aesop_cat
```
But `f_η'` and `f_μ'` are not in the original theorem and I get a new simpNF error:
```
#check CategoryTheory.monadToFunctor_mapIso_monad_iso_mk' /- Left-hand side does not simplify, when using the simp lemma on itself.
This usually means that it will never apply.
```
What do I do now?

Reply:
Hmm, thanks for debugging. It's really quite annoying that this new statement is now also flagged by the simp linter. I had expected that it would be a good simp lemma.
Please report in the porting note that you made this attempt, but that the simp linter was still not satisfied.

Reply:
Done.